runtime/src/bank.rs

```diff
@@ -1491,10 +1495,11 @@ impl Bank {
         // genesis needs stakes for all epochs up to the epoch implied by
         //  slot = 0 and genesis configuration
         {
-            let stakes = bank.stakes_cache.stakes();
+            let stakes = bank.stakes_cache.stakes().clone();
+            let stakes = Arc::new(StakesEnum::from(stakes));
             for epoch in 0..=bank.get_leader_schedule_epoch(bank.slot) {
                 bank.epoch_stakes
-                    .insert(epoch, EpochStakes::new(&stakes, epoch));
+                    .insert(epoch, EpochStakes::new(stakes.clone(), epoch));
```

Comment:
To modify an `Arc` you need to call [`Arc::get_mut`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.get_mut) which checks that there is only one reference, or call [`Arc::make_mut`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.make_mut) which clones the inner value if there are more than one reference.
So `Arc` by itself does not share mutations and even if `EpochStakes` are changed in the future to have some mutations, this is still safe.

Unless there is an interior mutablity pattern such as `Arc<RwLock<...>>` or some atomics inside, in which case we _do_ want to share mutations across references (e.g. an internal cache or something); or otherwise should not use interior mutability to begin with.

Reply:
> To modify an Arc you need to call [Arc::get_mut](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.get_mut) which checks that there is only one reference, or call [Arc::make_mut](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.make_mut) which clones the inner value if there are more than one reference.
So Arc by itself does not share mutations and even if EpochStakes are changed in the future to have some mutations, this is still safe.

I'm aware of this but thanks for the refresher! Looks like the only field with interior mutability that isn't arc wrapped is `staked_nodes` inside `VoteAccounts` so the worst that can happen is we recompute that more often. Seems fine to me